File: services/server/src/protocol/transport.py
import socket

import safe_socket

from .message import (
    _HEADER_SIZE,
    _LENGTH_FIELD_SIZE,
    _MAX_PAYLOAD_SIZE,
    _TYPE_FIELD_SIZE,
    Message,
)


def send_message(sock: socket.socket, message: Message) -> None:
    if len(message.payload) > _MAX_PAYLOAD_SIZE:
        raise RuntimeError(f"payload exceeds {_MAX_PAYLOAD_SIZE} bytes")

    header = get_header(message)
    safe_socket.send_all(sock, header + message.payload)


def recv_message(sock: socket.socket) -> Message:
    header = safe_socket.recv_all(sock, _HEADER_SIZE)

    message_type = header[0]
    payload_size = int.from_bytes(header[_TYPE_FIELD_SIZE:], "big")

    try:
        payload = safe_socket.recv_all(sock, payload_size)
    except RuntimeError:
        raise RuntimeError("connection closed mid-message")

    return Message(message_type, payload)


def get_header(message):
    return bytes([message.type]) + len(message.payload).to_bytes(
        _LENGTH_FIELD_SIZE, "big"
    )

File: services/client/src/protocol/transport.go
package protocol

import (
	"encoding/binary"
	"fmt"
	"io"

	"github.com/7574-sistemas-distribuidos/tp-nivelador/src/safe_socket"
)

func NewSendBuffer() []byte {
	initialLen := 0
	initialCapacity := _HEADER_SIZE + _MAX_PAYLOAD_SIZE
	return make([]byte, initialLen, initialCapacity)
}

func SendMessage(writer io.Writer, msg Message, buffer []byte) error {
	if len(msg.Payload) > _MAX_PAYLOAD_SIZE {
		return fmt.Errorf("payload exceeds %d bytes", _MAX_PAYLOAD_SIZE)
	}

	dataToSend := append(buffer[:0], byte(msg.Type))
	dataToSend = binary.BigEndian.AppendUint16(dataToSend, uint16(len(msg.Payload)))
	dataToSend = append(dataToSend, msg.Payload...)

	return safe_socket.SendAll(writer, dataToSend)
}

func RecvMessage(reader io.Reader) (Message, error) {
	header, err := safe_socket.RecvAll(reader, _HEADER_SIZE)
	if err != nil {
		return Message{}, err
	}

	msgType := MessageType(header[0])
	payloadSize := int(binary.BigEndian.Uint16(header[_TYPE_FIELD_SIZE:]))
	payload, err := safe_socket.RecvAll(reader, payloadSize)
	if err != nil {
		return Message{}, err
	}

	msg := NewMessage(msgType, payload)
	return msg, nil
}
